core/ledger/test/ledger_test.go

```diff
@@ -0,0 +1,1128 @@
+		})
+		// Block 1
+		It("creates, populates and finishes a batch", func() {
+			Expect(ledgerPtr.BeginTxBatch(1)).To(BeNil())
+			ledgerPtr.TxBegin("txUuid1")
+			Expect(ledgerPtr.SetState("chaincode1", "key1", []byte("value1B"))).To(BeNil())
+			Expect(ledgerPtr.SetState("chaincode2", "key2", []byte("value2B"))).To(BeNil())
+			Expect(ledgerPtr.SetState("chaincode3", "key3", []byte("value3B"))).To(BeNil())
+			ledgerPtr.TxFinished("txUuid1", true)
+		})
+		It("should commit the batch", func() {
+			uuid := util.GenerateUUID()
+			tx, err := protos.NewTransaction(protos.ChaincodeID{Path: "testUrl"}, uuid, "anyfunction", []string{"param1, param2"})
+			Expect(err).To(BeNil())
+			err = ledgerPtr.CommitTxBatch(1, []*protos.Transaction{tx}, nil, []byte("proof"))
+			Expect(err).To(BeNil())
+		})
+		It("should return committed state from batch 2", func() {
+			state, _ := ledgerPtr.GetState("chaincode1", "key1", true)
+			Expect(state).To(Equal([]byte("value1B")))
+			state, _ = ledgerPtr.GetState("chaincode2", "key2", true)
+			Expect(state).To(Equal([]byte("value2B")))
+			state, _ = ledgerPtr.GetState("chaincode3", "key3", true)
+			Expect(state).To(Equal([]byte("value3B")))
+		})
+		// Block 2
+		It("creates, populates and finishes a batch", func() {
+			Expect(ledgerPtr.BeginTxBatch(2)).To(BeNil())
+			ledgerPtr.TxBegin("txUuid1")
+			Expect(ledgerPtr.SetState("chaincode1", "key1", []byte("value1C"))).To(BeNil())
+			Expect(ledgerPtr.SetState("chaincode2", "key2", []byte("value2C"))).To(BeNil())
+			Expect(ledgerPtr.SetState("chaincode3", "key3", []byte("value3C"))).To(BeNil())
+			Expect(ledgerPtr.SetState("chaincode4", "key4", []byte("value4C"))).To(BeNil())
+			ledgerPtr.TxFinished("txUuid1", true)
+		})
+		It("should commit the batch", func() {
+			uuid := util.GenerateUUID()
+			tx, err := protos.NewTransaction(protos.ChaincodeID{Path: "testUrl"}, uuid, "anyfunction", []string{"param1, param2"})
+			Expect(err).To(BeNil())
+			err = ledgerPtr.CommitTxBatch(2, []*protos.Transaction{tx}, nil, []byte("proof"))
+			Expect(err).To(BeNil())
+		})
+		It("should return committed state from batch 3", func() {
+			state, _ := ledgerPtr.GetState("chaincode1", "key1", true)
+			Expect(state).To(Equal([]byte("value1C")))
+			state, _ = ledgerPtr.GetState("chaincode2", "key2", true)
+			Expect(state).To(Equal([]byte("value2C")))
+			state, _ = ledgerPtr.GetState("chaincode3", "key3", true)
+			Expect(state).To(Equal([]byte("value3C")))
+			state, _ = ledgerPtr.GetState("chaincode4", "key4", true)
+			Expect(state).To(Equal([]byte("value4C")))
+		})
+		// Roll backwards once
+		It("rolls backwards once without error", func() {
+			delta2, err := ledgerPtr.GetStateDelta(2)
+			Expect(err).To(BeNil())
+			delta2.RollBackwards = true
+			err = ledgerPtr.ApplyStateDelta(1, delta2)
+			Expect(err).To(BeNil())
+			err = ledgerPtr.CommitStateDelta(1)
+			Expect(err).To(BeNil())
+		})
+		PIt("should return committed state from batch 2 after rollback", func() {
```

Comment:
Note that this is the test that is currently skipped, which reveals that there is a leak from a previous database (which was deleted, no less) that can find its way into a subsequent database. Specifically, "chaincode4", "key4" returns a value of "value4" which AFAIK was put into a previous database at either line 205 or 258.